File: ios-app/Sources/RedCodeNetworking/APIClient.swift
import Foundation
import RedCodeCore

public protocol HTTPTransport: Sendable {
    func data(for request: URLRequest) async throws -> (Data, HTTPURLResponse)
}

public struct URLSessionHTTPTransport: HTTPTransport {
    private let session: URLSession

    public init(session: URLSession = .shared) {
        self.session = session
    }

    public func data(for request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw RedCodeError.network("响应不是 HTTPURLResponse")
        }
        return (data, httpResponse)
    }
}

public actor APIClient {
    private let environment: RedCodeEnvironment
    private let transport: any HTTPTransport
    private let encoder: JSONEncoder
    private let decoder: JSONDecoder

    public init(
        environment: RedCodeEnvironment,
        transport: any HTTPTransport = URLSessionHTTPTransport(),
        encoder: JSONEncoder = JSONEncoder(),
        decoder: JSONDecoder = JSONDecoder()
    ) {
        self.environment = environment
        self.transport = transport
        self.encoder = encoder
        self.decoder = decoder
    }

    public func get<Response: Decodable & Sendable>(
        _ endpoint: APIEndpoint,
        bearerToken: String? = nil,
        as responseType: Response.Type = Response.self
    ) async throws -> Response {
        try await send(endpoint, bearerToken: bearerToken, bodyData: nil, as: responseType)
    }

    public func post<Body: Encodable & Sendable, Response: Decodable & Sendable>(
        _ endpoint: APIEndpoint,
        body: Body,
        bearerToken: String? = nil,
        as responseType: Response.Type = Response.self
    ) async throws -> Response {
        let data = try encoder.encode(body)
        return try await send(endpoint, bearerToken: bearerToken, bodyData: data, as: responseType)
    }

    public func postNoResponse<Body: Encodable & Sendable>(
        _ endpoint: APIEndpoint,
        body: Body,
        bearerToken: String? = nil
    ) async throws {
        let data = try encoder.encode(body)
        try await sendNoResponse(endpoint, bearerToken: bearerToken, bodyData: data)
    }

    private func send<Response: Decodable & Sendable>(
        _ endpoint: APIEndpoint,
        bearerToken: String?,
        bodyData: Data?,
        as responseType: Response.Type
    ) async throws -> Response {
        let data = try await sendData(endpoint, bearerToken: bearerToken, bodyData: bodyData)
        do {
            return try decoder.decode(responseType, from: data)
        } catch {
            throw RedCodeError.network("响应解析失败")
        }
    }

    private func sendNoResponse(
        _ endpoint: APIEndpoint,
        bearerToken: String?,
        bodyData: Data?
    ) async throws {
        _ = try await sendData(endpoint, bearerToken: bearerToken, bodyData: bodyData)
    }

    private func sendData(
        _ endpoint: APIEndpoint,
        bearerToken: String?,
        bodyData: Data?
    ) async throws -> Data {
        let url = try endpoint.url(in: environment)
        var request = URLRequest(url: url)
        request.httpMethod = endpoint.method.rawValue
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let bodyData {
            request.httpBody = bodyData
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if let bearerToken, !bearerToken.isEmpty {
            request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        }

        let dataAndResponse: (Data, HTTPURLResponse)
        do {
            dataAndResponse = try await transport.data(for: request)
        } catch let error as RedCodeError {
            throw error
        } catch {
            throw RedCodeError.network(error.localizedDescription)
        }
        let (data, response) = dataAndResponse
        guard (200...299).contains(response.statusCode) else {
            throw RedCodeError.network(errorMessage(from: data) ?? "HTTP \(response.statusCode)")
        }
        return data
    }

    private func errorMessage(from data: Data) -> String? {
        guard !data.isEmpty else {
            return nil
        }
        guard
            let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return String(data: data, encoding: .utf8)
        }
        return object["message"] as? String
            ?? object["error"] as? String
            ?? object["detail"] as? String
    }
}

File: android-app/app/src/main/java/com/redcode/im/androidapp/network/APIClient.kt
package com.redcode.im.androidapp.network

import com.redcode.im.androidapp.core.config.RedCodeEnvironment
import kotlinx.serialization.SerializationException
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonPrimitive

class APIClient(
    @PublishedApi internal val environment: RedCodeEnvironment,
    @PublishedApi internal val transport: HttpTransport = JavaNetHttpTransport(),
    @PublishedApi internal val json: Json =
        Json {
            ignoreUnknownKeys = true
        },
) {
    suspend inline fun <reified Response : Any> get(
        endpoint: APIEndpoint,
        bearerToken: String? = null,
    ): Response =
        send(endpoint = endpoint, bearerToken = bearerToken, body = null)

    suspend inline fun <reified Body : Any, reified Response : Any> post(
        endpoint: APIEndpoint,
        body: Body,
        bearerToken: String? = null,
    ): Response =
        send(endpoint = endpoint, bearerToken = bearerToken, body = json.encodeToString(body))

    suspend inline fun <reified Body : Any, reified Response : Any> patch(
        endpoint: APIEndpoint,
        body: Body,
        bearerToken: String? = null,
    ): Response =
        send(endpoint = endpoint, bearerToken = bearerToken, body = json.encodeToString(body))

    suspend inline fun <reified Body : Any> postNoResponse(
        endpoint: APIEndpoint,
        body: Body,
        bearerToken: String? = null,
    ) {
        sendNoResponse(endpoint = endpoint, bearerToken = bearerToken, body = json.encodeToString(body))
    }

    suspend inline fun <reified Body : Any> patchNoResponse(
        endpoint: APIEndpoint,
        body: Body,
        bearerToken: String? = null,
    ) {
        sendNoResponse(endpoint = endpoint, bearerToken = bearerToken, body = json.encodeToString(body))
    }

    suspend fun sendNoResponse(
        endpoint: APIEndpoint,
        bearerToken: String? = null,
        body: String? = null,
    ) {
        executeRaw(endpoint = endpoint, bearerToken = bearerToken, body = body)
    }

    suspend inline fun <reified Response : Any> send(
        endpoint: APIEndpoint,
        bearerToken: String? = null,
        body: String? = null,
    ): Response {
        val responseBody = executeRaw(endpoint = endpoint, bearerToken = bearerToken, body = body)
        return try {
            json.decodeFromString(responseBody)
        } catch (error: SerializationException) {
            throw NetworkFailure(message = "响应解析失败", cause = error)
        }
    }

    @PublishedApi
    internal suspend fun executeRaw(
        endpoint: APIEndpoint,
        bearerToken: String? = null,
        body: String? = null,
    ): String {
        val headers = linkedMapOf("Accept" to "application/json")
        if (body != null) headers["Content-Type"] = "application/json"
        if (!bearerToken.isNullOrBlank()) headers["Authorization"] = "Bearer $bearerToken"

        val response =
            transport.execute(
                HttpRequest(
                    method = endpoint.method,
                    url = endpoint.url(environment),
                    headers = headers,
                    body = body,
                ),
            )
        if (response.statusCode !in 200..299) {
            throw NetworkFailure(
                statusCode = response.statusCode,
                message = extractErrorMessage(response.body) ?: "HTTP ${response.statusCode}",
            )
        }
        return response.body
    }

    fun extractErrorMessage(body: String): String? =
        runCatching {
            val element = json.parseToJsonElement(body)
            val obj = element as? JsonObject ?: return null
            listOf("message", "error", "detail")
                .firstNotNullOfOrNull { key ->
                    obj[key]?.jsonPrimitive?.content?.trim()?.takeIf { it.isNotBlank() }
                }
        }.getOrNull()
}
